In [1]:
"""
CS 6647
Chad E. McPherson
Dr. Huan
"""
#setup for future code blocks. Always run first.
import numpy as np

def get_future_state_probability_steps(P, S, num_days):
    steps = []

    steps.append(S)
    for i in range(num_days):
        steps.append(np.matmul(P, steps[i]))

    return steps

def print_steps(steps):
    for i in range(1, len(steps)):
        print('Probability on day ', i, ' is:', np.array_str(steps[i], precision=4))

In [2]:
# P is probability matrix from state diagram
P = [ [0.2, 0.1, 0.2], [0.6, 0.6, 0.7], [0.2, 0.3, 0.1] ]

# S is initial state (Resting)
S = [1, 0, 0]

# we want to know future state for day 15
steps = get_future_state_probability_steps(P, S, 15)
print_steps(steps)
print('Probability of exercise on day 15 is ', f'{steps[14][1]:.4f}')

Probability on day  1  is: [0.2 0.6 0.2]
Probability on day  2  is: [0.14 0.62 0.24]
Probability on day  3  is: [0.138 0.624 0.238]
Probability on day  4  is: [0.1376 0.6238 0.2386]
Probability on day  5  is: [0.1376 0.6239 0.2385]
Probability on day  6  is: [0.1376 0.6239 0.2385]
Probability on day  7  is: [0.1376 0.6239 0.2385]
Probability on day  8  is: [0.1376 0.6239 0.2385]
Probability on day  9  is: [0.1376 0.6239 0.2385]
Probability on day  10  is: [0.1376 0.6239 0.2385]
Probability on day  11  is: [0.1376 0.6239 0.2385]
Probability on day  12  is: [0.1376 0.6239 0.2385]
Probability on day  13  is: [0.1376 0.6239 0.2385]
Probability on day  14  is: [0.1376 0.6239 0.2385]
Probability on day  15  is: [0.1376 0.6239 0.2385]
Probability of exercise on day 15 is  0.6239


In [3]:
#3. the observation is the probability is the same on day 10,000 as it is on day 15.
# P is probability matrix from state diagram
P = [ [0.2, 0.1, 0.2], [0.6, 0.6, 0.7], [0.2, 0.3, 0.1] ]

# S is initial state (Resting)
S = [1, 0, 0]

# we want to know future state for day 10000
steps = get_future_state_probability_steps(P, S, 10000)

# If you really want to see 10,000 steps, uncomment this line
# print_steps(steps)
print('Probability of exercise on day 10,000 is ', f'{steps[9999][1]:.4f}')

Probability of exercise on day 10,000 is  0.6239


3. Ultimately, the observation is that on the fifth step, the system has become a steady state, so regardless, the exercise option will be at 0.6239 for infinity as its probability.  These two are identical because once the system got to its steady state, the numbers were guaranteed to be identical.